lorawan: adding NVM support for LoRaWAN and a reference SE implementation

GiulianoFranchetto opened the pull request:
For now, I have removed all the cryptographic calls from the implementation, as it was originally done using Nordic specific modules (mostly HUK and CryptoCell).

- [x] Need to abstract the crypto layer to not store the keys unencrytped on the flash
- [x] Various crypto must be managed, with great modularity. That will come from the fact that crypto API are very low-level stuff, and vendor dependent. I don't think you will have a Zephyr abstraction layer that will fit all use cases.
- [ ] Cryptos may come from west sub-project, and not zephyr tree

A more low-level implementation with key encryption using a daughter key can be found on the NCS repo: https://github.com/nrfconnect/sdk-zephyr/pull/678

Signed-off-by: Giuliano Franchetto <giuliano.franchetto@intellinium.com>

Mani-Sadhasivam reviewed (changes requested):
Can you please split this change into multiple commits for NVM and SE?

GiulianoFranchetto committed:
lorawan: adding settings SE for LoRaWAN

This commit proposes a reference implementation for the Secure-Element
needed to join any LoRaWAN network.

This implementation uses the SETTINGS backend.

More Secure Element implementation can be easily added either directly
in the source tree, or using a west module.

A west module should be the prefered implementation, as the SE
may require additionnal drivers or low-level security stuff
that is not natively part of Zephyr.

One example is the nRFx series which offers a LOT of security
modules (KMU, HUK, OTP, ITP ...).
A reference implemenation is proposed in the NCS fork of Zephyr
showing how to use some low-level security feature to make
the SE as secure as possible.

NOTE: the secure element implementation in this commit IS NOT
SECURED as it does not provide any clean way of storing the master
key. It is not recommended to be used in production (yet)

Signed-off-by: Giuliano Franchetto <giuliano.franchetto@intellinium.com>

GiulianoFranchetto committed:
lorawan: adding settings NVM for LoRaWAN

Adding a reference implementation of the Non-Volatile Memory module
needed to join any LoRaWAN network.

This NVM is based on the SETTINGS subsys to store all the required
key to join and communicate on a LoRaWAN network.

Without proper NVM, one may experience errors when using OTAA
to join the network, as the device may violate anti-replay
protection (depending on the version of LoRaWAN).

Signed-off-by: Giuliano Franchetto <giuliano.franchetto@intellinium.com>

JordanYates commented on subsys/lorawan/nvm/lorawan_nvm_settings.c:
```diff
@@ -0,0 +1,222 @@
+		mibReq.Param.Contexts->Crypto.JoinNonce);
+
+	/* Wait for LoRa to be idle */
+	while (LoRaMacStop() != LORAMAC_STATUS_OK) {
```
Why does the MAC layer need to be stopped and restarted here?

GiulianoFranchetto replied:
This is taken from the loramac-node [sample app](https://github.com/zephyrproject-rtos/loramac-node/blob/zephyr/src/apps/LoRaMac/common/NvmDataMgmt.c#L72).
I guess it is to prevent more IRQ to be fired which would also change the NVM values at the same time, inducing race condition.
However, I'm not an expert in the LoRaMAC layer, so that could be unnecessary.

JordanYates commented on subsys/lorawan/lorawan.c:
```diff
@@ -576,7 +576,7 @@ int lorawan_start(void)
 }
 
-static int lorawan_init(const struct device *dev)
+int lorawan_init(void)
```
What is the reasoning behind making applications call this function instead of making the priority user-configurable with Kconfig?

If you stick with this approach, the sample needs to be updated to actually call this.

GiulianoFranchetto replied:
Ok, corrected. That might be a change I did during testing that may not be necessary now.
I'll need to do the tests again just to be sure there is no issue with the initialization order of all the layers needed for the LoRaWAN stack.
The main issue I've found in the in the current implementation is that all the layers are deeply coupled, making the initialization a bit tricky

JordanYates commented on subsys/lorawan/lorawan.c:
```diff
@@ -591,6 +591,7 @@ static int lorawan_init(const struct device *dev)
 	macCallbacks.NvmDataChange = NULL;
 	macCallbacks.MacProcessNotify = OnMacProcessNotify;
 
+
```
Nit: newline

Mani-Sadhasivam commented on subsys/lorawan/nvm/Kconfig:
```diff
@@ -0,0 +1,22 @@
+	default LORAWAN_NVM_SETTINGS if SETTINGS
+	default LORAWAN_NVM_NONE
+
+config LORAWAN_NVM_NONE
```
I think we can get rid of this option.

GiulianoFranchetto replied:
What would happen if a user choose to use LoRaWAN without ```CONFIG_SETTINGS``` enabled ?
If we want to make LoRaWAN to only work if at least settings is enabled, should we ```select SETTINGS``` when enabling LoRaWAN ?

JordanYates replied:
Definitely not, I for example will need to write a custom backend for this as we don't use the settings API

Mani-Sadhasivam replied:
> Definitely not, I for example will need to write a custom backend for this as we don't use the settings API

Okay, in that case it makes sense!

Mani-Sadhasivam commented on subsys/lorawan/nvm/Kconfig:
```diff
@@ -0,0 +1,22 @@
+# LoRaWAN Non Volatile Memory configuration options
+
+# Copyright (c) 2021 Intellinium <giuliano.franchetto@intellinium.com>
```
2022 everywhere

GiulianoFranchetto replied:
Happy new year to all of you !

GiulianoFranchetto replied:
ps: done

Mani-Sadhasivam commented on subsys/lorawan/nvm/lorawan_nvm_settings.c:
```diff
@@ -0,0 +1,222 @@
+#include <settings/settings.h>
+#include "lorawan_nvm.h"
+#include <logging/log.h>
+#if defined(CONFIG_LORAWAN_SE_HAS_OWN_NVM)
```
Can you please make this commit as an independent PR? We should merge this one asap and backport to LTS.

GiulianoFranchetto replied:
working on it

Mani-Sadhasivam commented on subsys/lorawan/nvm/lorawan_nvm_settings.c:
```diff
@@ -0,0 +1,222 @@
+		k_sleep(K_MSEC(100));
+	}
+
+	SAVE_SETTINGS(nvm, LORAMAC_NVM_NOTIFY_FLAG_CRYPTO, Crypto)
```
Don't we need locking here as restore can happen in parallel?

GiulianoFranchetto replied:
```lorawan_nvm_data_restore()``` is only called when ```lorawan_init()``` is called.
As for ```lorawan_nvm_data_mgmt_event()```, I think the race condition is prevented thanks to [that](https://github.com/zephyrproject-rtos/zephyr/pull/40193#discussion_r785417057).

Mani-Sadhasivam reviewed (changes requested):
@GiulianoFranchetto Sorry taking time to respond to this PR. You should consider splitting these two commits into separate PRs so that we can merge NVM support first and backport to LTS.

JordanYates commented:
I should also note that the proposed API does not expose enough information for the application for it to decide when the Join EUI needs to be rotated (when DevNonce reaches UINT16_MAX)

GiulianoFranchetto commented on subsys/lorawan/nvm/lorawan_nvm_settings.c:
```diff
@@ -0,0 +1,222 @@
+	nvm_notify_flag = notifyFlags;
+
+	if (notifyFlags != LORAMAC_NVM_NOTIFY_FLAG_NONE) {
+		k_work_submit(&nvm_settings_worker);
```
ok, done

GiulianoFranchetto commented:
I agree that this is not exposed by the API. But is this really necessary ?
With the regulatory duty cycle that any LoRa endpoint must respect, I would guess that the DevNonce would reach UINT16_MAX after a very very very long time.

GiulianoFranchetto commented:
A draft PR with only NVM related changes has been created: #41903

JordanYates commented:
> I agree that this is not exposed by the API. But is this really necessary ?
With the regulatory duty cycle that any LoRa endpoint must respect, I would guess that the DevNonce would reach UINT16_MAX after a very very very long time.

That depends, the recommended maximum backoff averages at 30 minutes, which would overflow after 4 years. If the device is rebooting regularly however it could reach that value earlier.
